File: C++/PostorderTraversal/Solution2.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 
 
 
 /** This is the stack solution, it will modify the orignial tree, each time */
class Solution2 {
public:
    vector<int> postorderTraversal(TreeNode *root) {
        stack<TreeNode *> nodeStack;
        vector<int> result;
        if (root == nullptr) return result;
        nodeStack.push(root);
        TreeNode *curr;
        while (!nodeStack.empty()) {
            curr = nodeStack.top();
            if (curr->left != nullptr) {
                nodeStack.push(curr->left);
                curr->left = nullptr;
            } else if (curr->right != nullptr) {
                nodeStack.push(curr->right);
                curr->right = nullptr;
            } else {
                result.push_back(curr->val);
                nodeStack.pop();
            }
            
        }
        return result;
    }
};

File: Java/src/BinaryTreeInorderTraversal/Solution2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
                Stack<TreeNode> nodeStack = new Stack();
        List<Integer> result = new LinkedList<Integer>();

        
        TreeNode curr = root; 

        while (curr != null) {


            if (curr.left == null) {
                result.add(curr.val);
               
            } else {
                TreeNode left = curr.left;
                curr.left = null;
                nodeStack.push(curr);
                curr = left;
                continue;
            }

            if (curr.right != null) {
                nodeStack.push(curr.right);
            }
            
             curr = nodeStack.empty() ? null : nodeStack.pop();
        }
        
        return result;

    }

}


